frontend/src/pages/TaskMapPage.tsx

```diff
@@ -188,63 +105,141 @@ export const TaskMapPage = () => {
     const { sourceHandle, targetHandle } = data;
-    const roadmapId = currentRoadmap?.id;
     const type = TaskRelationType.Dependency;
 
     // Handles are in form 'from-{id}' and 'to-{id}', splitting required
     const from = Number(sourceHandle.split('-')[1]);
     const to = Number(targetHandle.split('-')[1]);
 
-    if (!roadmapId) return;
-    await dispatch(
-      roadmapsActions.addTaskRelation({
-        from,
-        to,
-        type,
-        roadmapId,
-      }),
-    );
+    await dispatch(roadmapsActions.addTaskRelation({ from, to, type }));
     dispatch(roadmapsActions.getRoadmaps());
   };
 
+  const addSynergyRelations = async (from: number, to: number[]) => {
+    const res = await dispatch(
+      roadmapsActions.addSynergyRelations({ from, to }),
+    );
+    if (roadmapsActions.addSynergyRelations.rejected.match(res))
+      throw new Error();
+  };
+
+  const onDragMoveOutside = async (
+    draggedTaskId: number,
+    copyList: GroupedRelation[],
+  ) => {
+    const newList = copyList
+      // remove dragged task from synergy list and
+      // all dependencies associated with the dragged task
+      .map(({ synergies, dependencies }) => ({
+        synergies: synergies.filter((num) => num !== draggedTaskId),
+        dependencies: dependencies.filter(
+          ({ from, to }) => from !== draggedTaskId && to !== draggedTaskId,
+        ),
+      }))
+      // remove entries where synergy list is empty
+      .filter(({ synergies }) => synergies.length);
+    newList.push({ synergies: [draggedTaskId], dependencies: [] });
+
+    setTaskRelations(newList);
+    await addSynergyRelations(draggedTaskId, []);
+  };
+
+  const onDragMoveToGroup = async (
+    source: DraggableLocation,
+    destination: DraggableLocation,
+    draggedTaskId: number,
+    copyList: GroupedRelation[],
+  ) => {
+    const destinationId = Number(destination.droppableId);
+
+    move()
+      .from(copyList[Number(source.droppableId)].synergies, source.index)
+      .to(copyList[destinationId].synergies, destination.index);
+
+    const newList = copyList
+      // remove entries where synergy list is empty
+      .filter(({ synergies }) => synergies.length)
+      // remove all dependencies associated with the dragged task
+      .map(({ synergies, dependencies }) => ({
+        synergies,
+        dependencies: dependencies.filter(
+          ({ from, to }) => from !== draggedTaskId && to !== draggedTaskId,
+        ),
+      }));
+
+    setTaskRelations(newList);
+    await addSynergyRelations(
+      draggedTaskId,
+      taskRelations[destinationId].synergies,
+    );
+  };
+
+  const onDragEnd = async (result: DropResult) => {
+    const { source, destination, draggableId } = result;
+    const draggedTaskId = Number(draggableId);
+    const backupList = copyRelationList(taskRelations);
+    const copyList = copyRelationList(taskRelations);
+
+    try {
+      if (!destination) await onDragMoveOutside(draggedTaskId, copyList);
```

Comment:
There are a few unnecessary copies, e.g. `onDragMoveOutside()` effectively does it's own copy with `map` & `filter`s,
and the `backupList` could be avoided if the possibly failing operations are done before setting the new state.

This was just an observation and no need to change this now

Reply:
Thanks for notifying about the multiple copies, I did not notice it
I wanted to temporarily set the new lists to the state so the animations look smooth (otherwise the task jumps around). Not sure how it's possible without an backup